pkg/generate/source/detector.go

```diff
@@ -0,0 +1,75 @@
+package source
+
+import (
+	"os"
+	"path/filepath"
+)
+
+type Info struct {
+	Platform string
+	Version  string
+}
+
+type DetectSource interface {
+	DetectSource(dir string) (*Info, bool)
+}
+
+type DetectorFunc func(dir string) (*Info, bool)
+
+type Detectors []DetectorFunc
+
+var DefaultDetectors = Detectors{
+	DetectRuby,
+	DetectJava,
+	DetectNodeJS,
+}
+
+type sourceDetector struct {
+	detectors []DetectorFunc
+}
+
+func (s Detectors) DetectSource(dir string) (*Info, bool) {
+	for _, d := range s {
+		if info, found := d(dir); found {
+			return info, true
+		}
+	}
+	return nil, false
+}
+
+func DetectRuby(dir string) (*Info, bool) {
```

Comment:
Shouldn't we have interface and detector factory? So we can add more detectors? Since all the Detect* function bellow have the same parameters?

```
type Detector interface {
  Detect(string) *Info,bool
}
```